### ISSUE TITLE
ajax-datatables-rails (0.2.0) with jquery-datatables-rails (2.1.10.0.2) - NoMethodError (undefined method `map' for nil:NilClass)

### ISSUE THREAD

**Issue by sushant-personal** [jbox-web/ajax-datatables-rails#32]
I am facing a server-side issue with datatable. I have given all the details in the following post is StackOverflow: http://stackoverflow.com/questions/25095611/rails4-ajax-datatables-rails-nomethoderror-undefined-method-map-for-nilnil

Any help will be great.

**Comment by antillas21** [comment]
Hi @sushant2675 

I see in the related StackOverflow post that you're using jQuery dataTables 1.9 syntax in your JS file... and that you're using the latest version for this gem (v. 0.2.0).

Version 0.2.0 is not compatible with jQuery dataTables 1.9... if you want to use that version of the jQuery plugin, you should use this gem's version 0.1.2. You can learn a little bit more about the changes [here](https://github.com/antillas21/ajax-datatables-rails/blob/master/CHANGELOG.md) and [here](https://github.com/antillas21/ajax-datatables-rails/tree/legacy).

Basically, changes relate to the params used to sort, filter and paginate records after you declare your `get_raw_records` method, as well as the JSON structure returned to feed to jQuery dataTables.

**Comment by sushant-personal** [comment]
Hi @antillas21 
Thanks for your reply. So as pointed by you, the syntax in JS file is only the issue. I changed the JS code to this:

jQuery(document).ready(function() {
    $('#dataentriesDatatable').dataTable({
        'bProcessing': true,
        'bServerSide': true,
        "deferRender": true,
        'sAjaxSource': $('#dataentriesDatatable').data('source'),
        'sPaginationType': 'bootstrap'
        // optional, if you want full pagination controls.
        // Check dataTables documentation to learn more about
        // available options.
    });
});

But  now I get the error: 

NoMethodError (undefined method `each_with_index' for nil:NilClass):
  app/controllers/dataentries_controller.rb:11:in`block (2 levels) in index'
  app/controllers/dataentries_controller.rb:9:in `index'

I am not able to figure out what I am doing wrong.

**Comment by antillas21** [comment]
@sushant2675 That error comes from [this method](https://github.com/antillas21/ajax-datatables-rails/blob/master/lib/ajax-datatables-rails/base.rb#L99-L105) right here.

Try the following:
- Add at least a string or text based database column to the array defined inside the `searchable_columns` method.
- Or, add column specification (in your JS) to mark them as non-searchable. See [this issue](https://github.com/antillas21/ajax-datatables-rails/issues/30#issuecomment-50666691) for a related approach. It's a thread in Spanish, but code will get the point accross.

**Comment by sushant-personal** [comment]
Hi @antillas21 
Well, I tried the following among others but still I am stuck now with:

NoMethodError (undefined method `[]' for nil:NilClass):
  app/controllers/dataentries_controller.rb:11:in`block (2 levels) in index'
  app/controllers/dataentries_controller.rb:9:in `index'

If you compare with @gipdiaz issue that you asked me to refer, his error generates from data method, but mine is from the controller itself.

I am generating a ActiveRecord relation and not an array, is that an issue? Well I also tried with to_a but I still get the same error.

  def index
    @dataentries = BookEntryAssignment.where(user_id: current_user.id).to_a
    respond_to do |format|
      format.html
      ### Following is line no. 11  generating error
      format.json { render json: DataentriesDatatable.new(view_context) }  
    end
  end

my JS file is:
jQuery(document).ready(function() {
    $('#dataentriesDatatable').dataTable({
        'bProcessing': true,
        'bServerSide': true,
        'sAjaxSource': $('#dataentriesDatatable').data('source'),
        'sPaginationType': 'bootstrap'
        // optional, if you want full pagination controls.
        // Check dataTables documentation to learn more about
        // available options.
    });
});

In my datatable.rb file, I have:
  def sortable_columns
    # list columns inside the Array in string dot notation.
    # Example: 'users.email'
    @sortable_columns ||= ['book_entry_assignments.id']
  end

  def searchable_columns
    # list columns inside the Array in string dot notation.
    # Example: 'users.email'
    @searchable_columns ||= ['book_entry_assignments.id']
  end

**Comment by antillas21** [comment]
@sushant2675 ,

Ok, could you create and link to a gist of the whole datatable.rb file?

In the meantime, what is the content of the `get_raw_records` method in this datatable.rb file?

Also:
- there is no need to declare a query to retrieve any records in the controller.rb file. So this:

``` ruby
def index
  @dataentries = BookEntryAssignment.where(user_id: current_user.id).to_a
  respond_to do |format|
    format.html
    ### Following is line no. 11 generating error
    format.json { render json: DataentriesDatatable.new(view_context) }
  end
end
```

should be this:

``` ruby
def index
  respond_to do |format|
    format.html
    format.json { render json: DataentriesDatatable.new(view_context) }
  end
end
```
- Record retrieval to display inside the dataTable **should happen** in the datatable.rb file, not the controller. So, if your `@dataentries` variable represents the data you want to display in the datatable, you would do something like this:

``` ruby
# controller.rb
def index
  respond_to do |format|
    format.html
    format.json { render json: DataentriesDatatable.new(view_context, { user: current_user }) }
  end
end

# datatable.rb
# lines defining methods
def get_raw_records
  BookEntryAssignment.where(user: options[:user])
  # or if you like using ids
  # BookEntryAssignment.where(user_id: options[:user].id)
end
```
- And yes, the result from the `get_raw_records` method **should be** an `ActiveRecord::Relation` object, and not an `Array`. Refer to this [important note](https://github.com/antillas21/ajax-datatables-rails#get-raw-records) to learn why.

**Comment by sushant-personal** [comment]
Hi @antillas21 
Thanks for your quick response. I am still getting the same error:

```
NoMethodError (undefined method `[]' for nil:NilClass):
  app/controllers/dataentries_controller.rb:11:in `block (2 levels) in index'
  app/controllers/dataentries_controller.rb:9:in `index'
```

My controller is:

```
class DataentriesController < ApplicationController
  before_filter :authenticate_user!
  #before_action :set_dataentry, only: [:show, :edit, :update, :destroy]

  # GET /dataentries
  # GET /dataentries.json
  def index
    #@dataentries = BookEntryAssignment.where(user_id: current_user.id)
    respond_to do |format|
      format.html
      format.json { render json: DataentriesDatatable.new(view_context, { user: current_user }) }
    end
  end

  # GET /dataentries/1
  # GET /dataentries/1.json
  def show
  end

  # GET /dataentries/new
  def new
    @dataentry = Dataentry.new
  end

  # GET /dataentries/1/edit
  def edit
  end

 end

```

My datatable.rb file is:

```
class DataentriesDatatable < AjaxDatatablesRails::Base
  # uncomment the appropriate paginator module,
  # depending on gems available in your project.
  include AjaxDatatablesRails::Extensions::Kaminari
  #def_delegators :@view


  def sortable_columns
    # list columns inside the Array in string dot notation.
    # Example: 'users.email'
    @sortable_columns ||= ['book_entry_assignments.id']
  end

  def searchable_columns
    # list columns inside the Array in string dot notation.
    # Example: 'users.email'
    @searchable_columns ||= ['book_entry_assignments.id']
  end

  private

  def data
    records.map do |record|
      [
          record.id,
          record.id,
          record.id,
          record.id,
          record.id,
          record.id,
          record.id
          #display_image(record, record.book.image_path, record.book.image_name),
          #record.book.name,
          #record.book.subject.name,
          #record.book.standard.name,
          #data_entry_status(record),
          #record.book.publish_status.titleize
          #action_buttons(record)
      # other attributes
      # comma separated list of the values for each cell of a table row
      # example: record.attribute,
      ]
    end
  end

  def get_raw_records
    # insert query here
    BookEntryAssignment.where(user: options[:user])
  end


  # ==== Insert 'presenter'-like methods below if necessary

end
```

I have made them very straight forward by trimming to minimum code, but still the error persists. I tried to debug and found that it generates error after the following line in datatable.rb:

```
@searchable_columns ||= ['book_entry_assignments.id']
```

My id column is of type uuid in postgresql 9.3. When I dug further, for id it expects to be Numeric but uuid is string so I think that's the problem. 

I tried again by changing the sortable_columns to:

```
  def searchable_columns
    # list columns inside the Array in string dot notation.
    # Example: 'users.email'
    @searchable_columns ||= ['book_entry_assignments.status']
  end
```

But still I get the same error. Status is an integer field in db but an enum defined in model:

```
class BookEntryAssignment < ActiveRecord::Base
  belongs_to :book
  belongs_to :user

  #BookEntryAssignment.reflect_on_all_associations(:belongs_to)

  has_paper_trail :only => [:book_id, :user_id, :status]
############ here is the enum ################
  enum status: [:pending, :"in-progress", :completed, :approved]
  after_initialize :set_default_status, :if => :new_record?

  def set_default_status
    self.status ||= :pending
  end

  validates :book_id, :user_id, :status, presence: true
  #, :if => lambda { |book| book.user.try(:admin) ==1 }
end
```

And I need search on status. There is no string field in BookEntryAssignment table so cannot test for search on string.

**Comment by antillas21** [comment]
@sushant2675 ,

The problem is, right now, we don't have support to perform searches on columns that are not String or Text based (at the database level), anything else: Integer, UUID, Float, Date, Time, etc. is not supported (for the time being).

This issue [has been reported here](https://github.com/antillas21/ajax-datatables-rails/issues/27) and there's a plan to tackle this in a future release. As always, all contributions are welcome, and if you'd like to tackle the problem and contribute to the gem, you're more than welcome to do so :smile:

**Comment by sushant-personal** [comment]
Thanks @antillas21 
For the support. I am relatively a newbie in ROR and have been solving small issues searching on the net. I shall be more than happy to know the intricacies and help.

**Comment by antillas21** [comment]
Nothing to thank for. I appreciate it when people use my gem and find scenarios I hadn't considered originally and report issues.

Well, invitation is open, if you'd like to contribute, please do so and open a pull request :smiley:

**Comment by sushant-personal** [comment]
Hi @antillas21 

Sorry, but this is stange that I get the same error:

```
NoMethodError (undefined method `[]' for nil:NilClass):
  app/controllers/dataentries_controller.rb:11:in `block (2 levels) in index'
  app/controllers/dataentries_controller.rb:9:in `index'
```

I have changed my query for the datatable: (rest is same)

```
SELECT "books".* FROM "books"  WHERE "books"."id" IN (SELECT "book_entry_assignments"."book_id" FROM "book_entry_assignments"  WHERE "book_entry_assignments"."user_id" = '<user_id>')
```

 and I get the following records from the DB:

```
"id";"name";"subject_id";"standard_id";"publisher";"printer";"isbn";"authors";"image_name";"image_path";"publish_status";"status_date";"user_id";"created_at";"updated_at"
"65ba346c-d30a-4ea2-9028-5d3cf75371d4";"Mathematics";"819aeb54-86b2-4c08-baf7-c61e8045b485";"b4fb5eb0-7b9b-42f1-a78b-80c8cefda62a";"";"";"";"";"Maths_Class_XI.jpg";"/public/book_images";0;"";"";"2014-07-29 00:10:53.527787";"2014-07-29 00:10:53.527787"
"84cbf371-700d-4b7d-964b-0e0bddd73823";"Mathematics";"819aeb54-86b2-4c08-baf7-c61e8045b485";"b4fb5eb0-7b9b-42f1-a78b-80c8cefda62a";"";"";"";"";"Maths_Class_XI.jpg";"/public/book_images";0;"";"";"2014-07-28 20:47:40.2496";"2014-07-30 17:12:02.227035"
"07440a91-4f62-40c9-956e-04189d378631";"Test 5";"cf3aea17-d404-4b69-9f4f-525e76a513db";"03c957fd-3149-4a1d-a3d3-dc483458a040";"";"";"";"";"default_no_book_image.gif";"/public/book_images";0;"";"";"2014-07-28 22:37:00.102463";"2014-07-28 22:37:00.102463"
```

The issue arises in:(name is varchar field in DB)

```
  def searchable_columns
    # list columns inside the Array in string dot notation.
    # Example: 'users.email'
    @searchable_columns ||= ['books.name']
  end
```

Digging deeper the exception is thrown here in conversions.rb, args[0] is nil:

```
    klass.send(:define_method, :to_s) do |*args|
      ############## in the following line args[0] is nil ###################
      if args[0].is_a?(Symbol)
        format = args[0]
        options = args[1] || {}

        self.to_formatted_s(format, options)
      else
        to_default_s(*args)
      end
    end
```

**Comment by antillas21** [comment]
Hi @sushant2675 ... I am sorry, but I don't recognize the code in your last block.

That is not part of this gem's codebase.

**Comment by sushant-personal** [comment]
Hi @antillas21 

The error is thrown here:

```
def aggregate_query
  conditions = searchable_columns.each_with_index.map do |column, index|
   ################# Here the params[:columns] is nil and ["#{index}"] is (1 element) "0" The error is thrown     here ###############
    value = params[:columns]["#{index}"][:search][:value]
    search_condition(column, value) unless value.blank?
  end
  conditions.compact.reduce(:and)
end
```

**Comment by antillas21** [comment]
@sushant2675 you mentioned you're using`jquery-datatables-rails` gem version 2.1.10.0.2. Could you please tell me if this version injects jQuery dataTables plugin version 1.9?

If it does, you need to use this gem (ajax-datatables-rails) version 0.1.x as the params hash keys are different, and also the resulting JSON structure is different.

I think that jquery-datatables-rails version 2.1.10.0.3 injects jQuery dataTables version 1.10 (which is compatible with this gem's latest release).

You can learn [more about versions here](https://github.com/antillas21/ajax-datatables-rails#versions).

**Comment by sushant-personal** [comment]
Hi @antillas21 

tell me if this version injects jQuery dataTables plugin version 1.9? ----->
How can I figure that out?

For the time being, I have reverted to only jquery datatables... it's working fine.

**Comment by ricardodovalle** [comment]
Hi @antillas21 and @sushant2675 , 

The last version of `jquery-datatables-rails`, to inject DataTables (1.9), was the gem version `1.12.2`, any other, it will inject DataTables 1.10.X.

The current version of gem `jquery-datatables-rails` is 2.2.1 with DataTables (1.10.2), it will works fine with ajax-datatables-rails, I have using at production. 

You can see here [CHANGELOG](https://github.com/rweng/jquery-datatables-rails/blob/master/CHANGELOG.md)

Thanks,
Great gem :+1: @antillas21

**Comment by panbanda** [comment]
Hey @antillas21, I am now getting the same error with params[:columns] is nil and ["#{index}"] is (1 element) "0" The error is thrown but I am not using jquery-datatables-rails, I am using angular, can we perhaps make the gem independent of url parameters?  I may do a patch if that would be helpful.  I'd like to access a url via json without passing parameters and it returning a standard response for 1.10+ datatables.  Thoughts?

**Comment by antillas21** [comment]
@jylinman you are more than welcome to do so :smile:

AFAIK, a base params hash is needed to perform the default data retrieval from the database through the `jquery.dataTable` plugin... but if this can be decoupled, I'm really open to suggestions.

Let me know if you need anything from me.
